Rabbitmq rearch

yadudoc opened the pull request:
I apologize in advance for the size of the PR. In the hope of making this PR not break the endpoint completely and to fix nuisances from ages past a bunch of changes have crept in here.

35ec8a0c4f17406fc0dfef6d2aaac24807620a2b : Renames `endpoint.py` -> `cli.py` and `endpoint_manage.py` to `endpoint.py`

We have 4 new classes:
`TaskQueueSubscriber`: This is an endpoint-side component designed to be launched off process that pipes task messages from the service over rabbit_mq into a multiprocessing.Queue.
`ResultQueuePublisher` : This class enables publishing results from the endpoint to the service over rabbit_MQ
`TaskQueuePublisher`: Service side component that publishes tasks to the endpoint over rabbit_mq
`ResultQueueSubscriber`: Service side component that listens for results from the endpoint over rabbit_mq.

Key bits that are already implemented and more-or-less tested:
* Disconnects raise and error in TaskQueueSubsciber and ResultQueuePublisher
* Once one TaskQueueSubscriber can be connected to an endpoint's rabbit_mq pipes preventing stealing by orphan EPs
* The subscribers all use RabbitMQ's `SelectConnection` which is fairly fast
* Longer duration connections are robust and do not timeout from missing heartbeats with the right components running asynchronously.
* ResultQueuePublisher publishes to a topic.

The bits that need work:
* No authz/n tests
* Tests assume there's a local rabbitMQ service with guest:guest as creds
* Limited/partial integration into the endpoint code replacing the existing ZMQ with the new code 
* The main limiting factor for ^ is the limited testing setup.

yadudoc committed:
endpoint.py is now renamed cli.py
endpoint_manager.py is renamed endpoint.py
EndpointManager is renamed Endpoint

yadudoc committed:
endpoint.py is now renamed cli.py
endpoint_manager.py is renamed endpoint.py
EndpointManager is renamed Endpoint

yadudoc committed:
More renaming

yadudoc committed:
Adds a new TaskQueueSubscriber, ResultQueueSubscriber, TaskQueuePublisher and ResultQueuePublisher

The TaskQueueSubscriber and ResultQueuePublisher together replaces the communication layer between the service and the endpoint. These
use RabbitMQ to replace current functionality built on ZMQ.

yadudoc committed:
Adds a new TaskQueueSubscriber, ResultQueueSubscriber, TaskQueuePublisher and ResultQueuePublisher

The TaskQueueSubscriber and ResultQueuePublisher together replaces the communication layer between the service and the endpoint. These
use RabbitMQ to replace current functionality built on ZMQ.

yadudoc committed:
Adding design notes on RabbitMQ

yadudoc committed:
Adding design notes on RabbitMQ

yadudoc committed:
Adding tests for the RabbitMQ components and some minimal mocking of the service<-->endpoint pipes.

yadudoc committed:
Adding tests for the RabbitMQ components and some minimal mocking of the service<-->endpoint pipes.

yadudoc committed:
Minor doc fix

yadudoc committed:
Minor doc fix

yadudoc commented:
This is related to [sc-11441], [sc-11460], and [sc-11449] although incomplete.

sirosen commented:
My first reaction -- I haven't started reading yet -- is that we should cherry-pick 35ec8a0 and ccbf171 into a new branch, get that reviewed and merged, and then rebase the remaining commits. That way, we won't have renaming-related noise in the review of a major functional change.
(The mechanism does not need to be cherry-pick. A PR from ccbf171f would do, as would any of a number of other similar approaches.)

sirosen commented on funcx_endpoint/funcx_endpoint/endpoint/rabbit_mq/result_queue_publisher.py:
```diff
@@ -0,0 +1,108 @@
+        self._channel = None
+        self._stopping = False
+
+    def connect(self) -> pika.BlockingConnection:
```
Rather than having `ResultQueuePublisher` responsible for constructing its own connections, I think it would be better to make `pika.BlockingConnection`. That would eliminate any need for `pika_conn_params`, so it simplifies. Plus it's more flexible: e.g. a mock connection could be passed for testing.

The same holds for the other classes which are wrapping connections. e.g. `ResultQueueSubscriber` could take a `pika.SelectConnection`.

sirosen commented on funcx_endpoint/funcx_endpoint/endpoint/rabbit_mq/result_queue_publisher.py:
```diff
@@ -0,0 +1,108 @@
+    def close(self):
+        self._channel.close()
+        self._conn.close()
+        self._is_connected = False
```
I'm not familiar with the pika/rabbitmq semantics for a channel. Do we need to close the channel when we are done using it?

If this is required, it might be better to make this behave as a context manager by defining `__enter__` and `__exit__`.

sirosen commented on funcx_endpoint/funcx_endpoint/endpoint/rabbit_mq/result_queue_publisher.py:
```diff
@@ -0,0 +1,108 @@
+        self.params.blocked_connection_timeout = 60
+        self.exchange = exchange
+        self.exchange_type = "topic"
+        self._is_connected = False
```
I don't see `_is_connected` being used anywhere.
Could we instead use

```python
@property
def is_connected(self) -> bool:
    return self._channel is not None
```
?

The fewer explicit attributes we're trying to track and keep in sync, the less chance we have of subtle bugs from them getting out of sync.

sirosen commented on funcx_endpoint/funcx_endpoint/endpoint/rabbit_mq/result_queue_publisher.py:
```diff
@@ -0,0 +1,108 @@
+                time.sleep(3 ** retry_count)  # Hacky exponential back-off
+                try:
+                    self.connect()
+                except Exception:
```
Do we have explicit, known, conditions under which a connection would be rejected that we want to retry?
`except Exception` is broad, so it's better to avoid it when possible.

sirosen commented on funcx_endpoint/funcx_endpoint/endpoint/rabbit_mq/result_queue_publisher.py:
```diff
@@ -0,0 +1,108 @@
+        self._is_connected = False
+
+    def _queue_purge(self):
+        """This method is *ONLY* for testing. This should not work in production"""
```
In that case, does it belong here? Or should tests simply grab `_channel` and do this action explicitly?

sirosen commented on funcx_endpoint/funcx_endpoint/endpoint/rabbit_mq/result_queue_subscriber.py:
```diff
@@ -0,0 +1,316 @@
+    def _on_open_failed(self, *args):
+        logger.warning(f"Connection failed to open : {args}")
+
+    def _on_connection_closed(self, connection, exception):
```
Seeing this method and how it's used to construct the `SelectConnection` is driving me more strongly to think that we should isolate creation and management of the connection objects into a dedicated set of helpers. These are details we should be working out how to share between the different connection classes.

For example,
```python
def on_connection_closed(connection: pika.Connection, exception: Exception) -> None:
    if isinstance(connection, pika.SelectConnection):
        _on_select_connection_closed(connection, exception)
    elif isinstance(connection, pika.BlockingConnection):
        _on_blocking_connection_closed(connection, exception)
    ...
```

sirosen reviewed (commented):
The main thing I'm trying to understand as I read the various new classes is whether or not there's a valid use-case for having an object which is not connected, or which is missing parameters needed to run.

For example, `ResultQueueSubscriber` has `__init__`, `connect`, and `run`.
I think `__init__` vs `run` makes sense as a separation. But can `connect` be eliminated by having a connection passed in from the outside? That would eliminate the possibility of either calling `ResultQueueSubscriber(...).run()` or `x = ResultQueueSubscriber(...); x.connect(); x.run()`.
Removing these variations simplifies. If `connect()` is meant as an internal method, let's rename to `_connect` to clarify that.

Similarly, are the parameters to `run` not possible to pass to `__init__` instead, and make `ResultQueueSubscriber.run` take no arguments? That eliminates the possibility of a `ResultQueueSubscriber` where `external_queue` is not set. Having attributes which are set for the first time outside of `__init__` is very confusing. At the very least it ought to be initialized as `None` in `__init__`. But maybe having it as a parameter to `run` is not necessary?

yadudoc committed:
* TaskQueueSubscriber now inherits from multiprocessing.Process, and methods have been updated to match.
    - An external_callback can now be supplied for testing
    - Updated docstrings
    - New event watcher to shutdown on setting kill_event
* TaskQueuePublisher has some minor interface updates
* ResultQueueSubscriber has interface updates matching TaskQueueSubscriber

yadudoc committed:
* TaskQueueSubscriber now inherits from multiprocessing.Process, and methods have been updated to match.
    - An external_callback can now be supplied for testing
    - Updated docstrings
    - New event watcher to shutdown on setting kill_event
* TaskQueuePublisher has some minor interface updates
* ResultQueueSubscriber has interface updates matching TaskQueueSubscriber

yadudoc committed:
Updating tests

yadudoc committed:
Updating tests

yadudoc committed:
Partial changes, need amend

yadudoc committed:
Partial changes, need amend

yadudoc committed:
Updating interchange to use process based TaskQueueSubscriber

yadudoc committed:
Updating interchange to use process based TaskQueueSubscriber

yadudoc committed:
Fix ResultQueueSubscriber _connect call

yadudoc committed:
Fix ResultQueueSubscriber _connect call

yadudoc committed:
Adding tests for endpoint registration and complete EP end-to-end

yadudoc committed:
Adding tests for endpoint registration and complete EP end-to-end

yadudoc committed:
Merge branch 'rabbitmq_rearch' of github.com:funcx-faas/funcX into rabbitmq_rearch

yadudoc committed:
Explicitly make worker_logs dir
Minor json load field name fix endpoint_id -> endpoint_uuid

sirosen commented:
@yadudoc, should we close this? (Not delete the branch, just close the draft PR). With the rename already merged and #701 seeming to supersede a chunk of this, I'm not clear on the purpose of this PR.

yadudoc commented:
Closing this since this is very stale and #701 and #703 have split out the changes and improved on it.
